### PR TITLE
Automatic GitHub Composer Build Testing, PHP 7.4.9 Compliance, Logging Update

### PR DESCRIPTION
**Pull request by timothygwebb**

### CODE REVIEW

**Comment by fordnox** [comment]
@timothygwebb need to rebase, too many commits

**Comment by timothygwebb** [comment]
@fordnox I will get latest and update with my changes then recommit. I believe that a new release will need to be completed after my commit is merged into the master branch because I will be updating all composer packages along with the changes to be 7.4.9 compliant and the build testing for composer.json and composer.lock files.

**Comment by S4-Hosting** [comment]
I think that there definitely needs to be a new numbered release once all of the current pull requests are merged. It's a big jump to be compatible with current PHP versions and surely deserves a significant version number. At the same time it should be updated on the BB site to solve #613

**Comment by timothygwebb** [comment]
Resolved Issues:

Updated Composer Packages.

Deprecated: Function get_magic_quotes_gpc() is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-load.php on line 141 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-modules/Theme/Service.php on line 279 

Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-modules/Theme/Service.php on line 279

Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 209 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 211 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 505 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 505 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 512 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 546 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 546 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 561 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 562 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 661 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 671 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 671 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 680 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 681 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 1068 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/MarkdownExtra.php on line 1463 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/Markdown.php on line 948 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/Markdown.php on line 1349 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/Markdown.php on line 1374 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/Markdown.php on line 1785 Deprecated: Array and string offset access syntax with curly braces is deprecated in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-vendor/michelf/php-markdown/Michelf/Markdown.php on line 1787 

[31-Aug-2020 15:19:33 UTC] 2 count(): Parameter must be an array or an object that implements Countable /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-library/Server/Manager/Plesk.php 602
[31-Aug-2020 15:19:33 UTC] PHP Warning:  count(): Parameter must be an array or an object that implements Countable in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-library/Server/Manager/Plesk.php on line 602
[31-Aug-2020 15:19:36 UTC] Connection to server failed 0
[31-Aug-2020 15:20:02 UTC] PHP Parse error:  syntax error, unexpected 'class' (T_CLASS), expecting identifier (T_STRING) or variable (T_VARIABLE) or '{' or '$' in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-di.php on line 153
[31-Aug-2020 15:24:54 UTC] 8 Trying to access array offset on value of type null /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-library/Box/Log.php 123
[31-Aug-2020 15:24:54 UTC] PHP Notice:  Trying to access array offset on value of type null in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-library/Box/Log.php on line 123
[31-Aug-2020 15:24:54 UTC] 8 Trying to access array offset on value of type null /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-library/Box/Log.php 123
[31-Aug-2020 15:24:54 UTC] PHP Notice:  Trying to access array offset on value of type null in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-library/Box/Log.php on line 123
[31-Aug-2020 15:25:01 UTC] PHP Parse error:  syntax error, unexpected 'class' (T_CLASS), expecting identifier (T_STRING) or variable (T_VARIABLE) or '{' or '$' in /var/www/vhosts/webbhostingservices.com/httpdocs/boxbilling/bb-di.php on line 153



Common subdirectories: boxbilling/bb-data and boxbilling.bak/bb-data
diff boxbilling/bb-di.php boxbilling.bak/bb-di.php
138c138
< $di['twig'] = $di->factory(function () use ($di) {
---
> $di['twig'] = function () use ($di) {
166c166
< });
---
> };
Common subdirectories: boxbilling/bb-library and boxbilling.bak/bb-library
diff boxbilling/bb-load.php boxbilling.bak/bb-load.php
141,142c141,142
< if (function_exists('get_magic_quotes_gpc') && get_magic_quotes_gpc()) {
<     // Create lamba style unescaping function (for portability)
---
> //if (function_exists('get_magic_quotes_gpc') && get_magic_quotes_gpc()) {
> // Create lamba style unescaping function (for portability)
144,153c144,152
< 
<     /* create_function deprecated in PHP 7.2
<     $unescape_function = (empty($quotes_sybase) || $quotes_sybase === 'off') ? 'stripslashes($value)' : 'str_replace("\'\'","\'",$value)';
<     $stripslashes_deep = create_function('&$value, $fn', '
<         if (is_string($value)) {
<             $value = ' . $unescape_function . ';
<         } else if (is_array($value)) {
<             foreach ($value as &$v) $fn($v, $fn);
<         }
<     ');  */
---
> //    /* create_function deprecated in PHP 7.2
> //    $unescape_function = (empty($quotes_sybase) || $quotes_sybase === 'off') ? 'stripslashes($value)' : 'str_replace("\'\'","\'",$value)';
> //    $stripslashes_deep = create_function('&$value, $fn', '
> //        if (is_string($value)) {
> //            $value = ' . $unescape_function . ';
> //        } else if (is_array($value)) {
> //            foreach ($value as &$v) $fn($v, $fn);
> //        }
> //  ');  */
169c168
< }
---
> 
Common subdirectories: boxbilling/bb-locale and boxbilling.bak/bb-locale
Common subdirectories: boxbilling/bb-modules and boxbilling.bak/bb-modules
Common subdirectories: boxbilling/bb-themes and boxbilling.bak/bb-themes
diff boxbilling/bb-update.php boxbilling.bak/bb-update.php
43a44,76
> 
>  * Version 4.12.1
>  */
> class BBPatch_20 extends BBPatchAbstract
> {
>     public function patch()
>     {
>         $q="ALTER TABLE product_payment ADD metered_price DECIMAL(18, 2) DEFAULT '0.00';";
>         $this->execSql($q);
>         $q="ALTER TABLE product_payment ADD metered_setup_price DECIMAL(18,2) DEFAULT '0.00';";
>         $this->execSql($q);
>         $q = "CREATE TABLE iF NOT EXISTS metered_usage
>                 (
>                     id bigint(20) NOT NULL AUTO_INCREMENT PRIMARY KEY,
>                     client_id bigint(20) NOT NULL,
>                     order_id bigint(20) NOT NULL,
>                     plan_id bigint(20) NOT NULL,
>                     product_id bigint(20) NOT NULL,
>                     invoice_id bigint(20) NOT NULL DEFAULT '0',
>                     quantity int NOT NULL COMMENT 'Duration in seconds',
>                     price double(18,8) NOT NULL COMMENT 'Price for 1 duration item',
>                     created_at DATETIME NOT NULL,
>                     stopped_at DATETIME,
>                     KEY `order_id` (`order_id`)
>                 ) ENGINE=MyISAM  DEFAULT CHARSET=utf8 AUTO_INCREMENT=1 ;";
>         $this->execSql($q);
> 
>     }
> 
> }
> /**
>  * Version X.XX.X
> =======
150a184
> 
158a193,195
> }
> 
> 
Common subdirectories: boxbilling/bb-uploads and boxbilling.bak/bb-uploads
Common subdirectories: boxbilling/bb-vendor and boxbilling.bak/bb-vendor
Common subdirectories: boxbilling/bin and boxbilling.bak/bin
diff boxbilling/composer.lock boxbilling.bak/composer.lock
1100c1100
<             "version": "v4.9.1",
---
>             "version": "v4.9.0",
1104c1104
<                 "reference": "88e519766fc58bd46b8265561fb79b54e2e00b28"
---
>                 "reference": "aaee038b912e567780949787d5fe1977be11a778"
1108,1109c1108,1109
<                 "url": "https://api.github.com/repos/nikic/PHP-Parser/zipball/88e519766fc58bd46b8265561fb79b54e2e00b28",
<                 "reference": "88e519766fc58bd46b8265561fb79b54e2e00b28",
---
>                 "url": "https://api.github.com/repos/nikic/PHP-Parser/zipball/aaee038b912e567780949787d5fe1977be11a778",
>                 "reference": "aaee038b912e567780949787d5fe1977be11a778",
1148c1148
<             "time": "2020-08-30T16:15:20+00:00"
---
>             "time": "2020-08-18T19:48:01+00:00"
1464c1464
<             "version": "9.1.6",
---
>             "version": "9.1.4",
1468c1468
<                 "reference": "d25b24b1cd14772bde4d75daeb393dc17db9f6e6"
---
>                 "reference": "4422fca28c3634e2de8c7c373af97a104dd1a45f"
1472,1473c1472,1473
<                 "url": "https://api.github.com/repos/sebastianbergmann/php-code-coverage/zipball/d25b24b1cd14772bde4d75daeb393dc17db9f6e6",
<                 "reference": "d25b24b1cd14772bde4d75daeb393dc17db9f6e6",
---
>                 "url": "https://api.github.com/repos/sebastianbergmann/php-code-coverage/zipball/4422fca28c3634e2de8c7c373af97a104dd1a45f",
>                 "reference": "4422fca28c3634e2de8c7c373af97a104dd1a45f",
1533c1533
<             "time": "2020-08-31T06:31:46+00:00"
---
>             "time": "2020-08-13T15:04:53+00:00"
1762c1762
<             "version": "9.3.8",
---
>             "version": "9.3.7",
1766c1766
<                 "reference": "93d78d8e2a06393a0d0c1ead6fe9984f1af1f88c"
---
>                 "reference": "c638a0cac77347980352485912de48c99b42ad00"
1770,1771c1770,1771
<                 "url": "https://api.github.com/repos/sebastianbergmann/phpunit/zipball/93d78d8e2a06393a0d0c1ead6fe9984f1af1f88c",
<                 "reference": "93d78d8e2a06393a0d0c1ead6fe9984f1af1f88c",
---
>                 "url": "https://api.github.com/repos/sebastianbergmann/phpunit/zipball/c638a0cac77347980352485912de48c99b42ad00",
>                 "reference": "c638a0cac77347980352485912de48c99b42ad00",
1787c1787
<                 "phpunit/php-code-coverage": "^9.1.5",
---
>                 "phpunit/php-code-coverage": "^9.1.1",
1792d1791
<                 "sebastian/cli-parser": "^1.0",
1857,1909c1856
<             "time": "2020-08-27T06:30:58+00:00"
<         },
<         {
<             "name": "sebastian/cli-parser",
<             "version": "1.0.0",
<             "source": {
<                 "type": "git",
<                 "url": "https://github.com/sebastianbergmann/cli-parser.git",
<                 "reference": "2a4a38c56e62f7295bedb8b1b7439ad523d4ea82"
<             },
<             "dist": {
<                 "type": "zip",
<                 "url": "https://api.github.com/repos/sebastianbergmann/cli-parser/zipball/2a4a38c56e62f7295bedb8b1b7439ad523d4ea82",
<                 "reference": "2a4a38c56e62f7295bedb8b1b7439ad523d4ea82",
<                 "shasum": ""
<             },
<             "require": {
<                 "php": "^7.3 || ^8.0"
<             },
<             "require-dev": {
<                 "phpunit/phpunit": "^9.3"
<             },
<             "type": "library",
<             "extra": {
<                 "branch-alias": {
<                     "dev-master": "1.0-dev"
<                 }
<             },
<             "autoload": {
<                 "classmap": [
<                     "src/"
<                 ]
<             },
<             "notification-url": "https://packagist.org/downloads/",
<             "license": [
<                 "BSD-3-Clause"
<             ],
<             "authors": [
<                 {
<                     "name": "Sebastian Bergmann",
<                     "email": "sebastian@phpunit.de",
<                     "role": "lead"
<                 }
<             ],
<             "description": "Library for parsing CLI options",
<             "homepage": "https://github.com/sebastianbergmann/cli-parser",
<             "funding": [
<                 {
<                     "url": "https://github.com/sebastianbergmann",
<                     "type": "github"
<                 }
<             ],
<             "time": "2020-08-12T10:49:21+00:00"
---
>             "time": "2020-08-11T15:36:12+00:00"
Common subdirectories: boxbilling/data and boxbilling.bak/data
Common subdirectories: boxbilling/.git and boxbilling.bak/.git
Common subdirectories: boxbilling/.github and boxbilling.bak/.github
Only in boxbilling.bak: .htaccess
Common subdirectories: boxbilling/install and boxbilling.bak/install
diff boxbilling/phpunit.xml.dist boxbilling.bak/phpunit.xml.dist
2,34c2,34
< <phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" backupGlobals="false" backupStaticAttributes="false" colors="true" convertErrorsToExceptions="true" convertNoticesToExceptions="true" convertWarningsToExceptions="true" processIsolation="false" stopOnFailure="false" beStrictAboutChangesToGlobalState="true" bootstrap="tests/bootstrap.php" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd">
<   <coverage processUncoveredFiles="true">
<     <include>
<       <directory suffix=".php">src</directory>
<     </include>
<     <exclude>
<       <file>src/rb.php</file>
<       <file>src/bb-modules/Spamchecker/recaptchalib.php</file>
<       <file>src/bb-modules/Spamchecker/akismet.curl.class.php</file>
<       <file>src/bb-library/Box/Zip.php</file>
<       <directory>src/bb-vendor</directory>
<       <directory>src/bb-data</directory>
<       <directory>src/bb-library/Registrar/includes</directory>
<       <directory>src/bb-modules/Filemanager/src</directory>
<       <directory>src/bb-modules/Servicecentovacast/ccapiclient</directory>
<       <directory>src/bb-library/php-gettext</directory>
<       <directory>src/bb-themes</directory>
<       <directory>src/install</directory>
<     </exclude>
<   </coverage>
<   <testsuites>
<     <testsuite name="Library">
<       <directory>tests/bb-library</directory>
<     </testsuite>
<     <testsuite name="Integration">
<       <directory>tests/integration</directory>
<     </testsuite>
<   </testsuites>
<   <php>
<     <ini name="display_errors" value="On"/>
<     <ini name="display_startup_errors" value="On"/>
<   </php>
< </phpunit>
---
> 
> <phpunit backupGlobals="false"
>          backupStaticAttributes="false"
>          colors="true"
>          convertErrorsToExceptions="true"
>          convertNoticesToExceptions="true"
>          convertWarningsToExceptions="true"
>          processIsolation="false"
>          stopOnFailure="false"
>          syntaxCheck="false"
>          beStrictAboutChangesToGlobalState="true"
>          bootstrap="tests/bootstrap.php"
>         >
>     <filter>
>         <whitelist excludeVendorFiles="true">
>             <directory suffix=".php">src</directory>
>             <exclude>
>                 <file>src/rb.php</file>
>                 <file>src/bb-modules/Spamchecker/recaptchalib.php</file>
>                 <file>src/bb-modules/Spamchecker/akismet.curl.class.php</file>
>                 <file>src/bb-library/Box/Zip.php</file>
>                 <directory>src/bb-vendor</directory>
>                 <directory>src/bb-data</directory>
>                 <directory>src/bb-library/Registrar/includes</directory>
>                 <directory>src/bb-modules/Filemanager/src</directory>
>                 <directory>src/bb-modules/Servicecentovacast/ccapiclient</directory>
>                 <directory>src/bb-library/php-gettext</directory>
>                 <directory>src/bb-themes</directory>
>                 <directory>src/install</directory>
>             </exclude>
>         </whitelist>
>     </filter>
> </phpunit>
\ No newline at end of file
Common subdirectories: boxbilling/puphpet and boxbilling.bak/puphpet
Common subdirectories: boxbilling/src and boxbilling.bak/src
Common subdirectories: boxbilling/tests and boxbilling.bak/tests

File: README.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ Open Source billing software
 Requirements
 ================================================================================
 
-* PHP 7.4.8
+* PHP 7.2
```
Comment (timothygwebb) [inline README.md:10]: Master branch overwrote my update for the PHP version I will need to modify.

**Pull request closed without merging**